### DISCUSSION TITLE
Navigating Big Projects

### DISCUSSION THREAD

**Discussion by dmrvb** [twinbasic/twinbasic#293] [Ideas]
I plan to put some rather big projects into TwinBasic.
Putting everything into one big Twin file makes the code very difficult to navigate.
Splitting things into separate CLS, BAS (or Twin)  files works well and keeps my current VB6 structure, and I can use sub folders which is rather nice.
 
<img src="https://user-images.githubusercontent.com/83877219/126867419-f4afe3a2-f288-4bfa-a657-a3e1339ccc5d.png" width="200">

Twin files are good because they can contain more than one class, and in particular can hold both a class and any associated public definitions in a module. Maybe there is no need for the BAS and CLS files?

It would be good to have user selectable icons for the CLS, BAS, and TWIN files, maybe these could be the standard TB icon with different colour schemes? 

Even with separate files a big class with lots of big methods is still very clumsy to navigate. There are two things I would like see:

1 In VB6 the editor window can show either the full class/module or just the current method/sub/function and its really quick to switch between these modes, something similar in TB would really help.

2 The outline or breadcrumbs is the way to navigate the code but these both show Much too much info...every public and private variable is shown, plus every API call definition (the API definitions even use the same icon as the methods). An option to show only "items containing code" (methods/functions/properties etc) would be really useful.

I am new to VScode so maybe this functionality exists but I can't find it.

**Comment by WaynePhillipsEA** [comment]
I do like the idea of having the tB icon different colours for CLS/BAS files.

For 1) VS Code includes a command for 'Fold all regions except selected' which might help you somewhat.  (default key binding is CTRL and K followed by CTRL and -)

For 2) I find the breadcrumb bar is better for navigation, as it offers separate levels of navigation (for folders > files > component > procedures > variables).  Personally I prefer to hide the file path from the breadcrumb bar to make it cleaner.  To do that, there is an option in the VS code preferences called 'Breadcrumbs: File Path' and set it to OFF.

Something that I also find very useful, and didn't notice until recently, was that in most VS code areas you can just start typing to filter lists.  e.g. drop down a breadcrumb and just start typing the name of the item you're looking for.

**Reply by WaynePhillipsEA** [reply]
<img width="194" alt="tb_Icon_Colours" src="https://user-images.githubusercontent.com/25738109/126897032-ff608477-893b-47ef-b1ba-ccc1dbd16fb2.png">

Colours now available in v0.10.3472

**Reply by dmrvb** [reply]
Thank you. These look good.
Could we have a few more please? Making them user assignable would be great but otherwise can we have two different colours for the .TWIN's depending upon whether they are added as class or module?

**Reply by WaynePhillipsEA** [reply]
Really this would need to be based on the .twin file content (rather than just the file extension).  It can be done, but things are not quite setup for that yet.
